app/services/carto/visualization_backup_service.rb

```
@@ -0,0 +1,39 @@
+require_dependency 'carto/visualizations_export_service_2'
+
+module Carto
+  module VisualizationBackupService
+    include Carto::VisualizationsExportService2Exporter
+    include Carto::VisualizationsExportService2Importer
+
+    def create_visualization_backup(visualization:, category:, with_mapcaps: true, with_password: true)
+      export_json = export_visualization_json_hash(
+        visualization.id,
+        visualization.user,
+        with_mapcaps: with_mapcaps,
+        with_password: with_password
+      )
+      Carto::VisualizationBackup.create!(
+        user_id: visualization.user.id,
+        visualization_id: visualization.id,
+        category: category,
+        export: export_json
+      )
+    rescue StandardError => exception
+      # The backup should not break the flow
+      CartoDB::Logger.error(
+        message: 'Error creating a visualization backup',
+        exception: exception,
+        visualization: visualization
+      )
+    end
+
+    def restore_visualization_backup(visualization_backup_id)
+      visualization_backup = Carto::VisualizationBackup.find(visualization_backup_id)
+      user = Carto::User.find(visualization_backup.user_id)
+
+      json_export = JSON.dump(visualization_backup.export)
+      visualization_json = build_visualization_from_json_export(json_export)
+      Carto::VisualizationsExportPersistenceService.new.save_import(user, visualization_json, full_restore: true)
```

Maybe raise an error if a visualization with the same ID as indicated in the backup already exists? If not, it will fail here with an uglier error.

---

Same as previous one